Make named fenics lib

ravenjoad opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Before, the dolfin (and supporting) packages were defined in a `let`-binding, for the `pythonPackages.fenics` package.  There are some other packages that use the library that dolfin provides (`dolfin.h` and others). For example, there are several packages for octave that rely on the dolfin/fenics package being available for use. 


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I should say that right now `python37Packages.dolfin` and `python39Packages.dolfin` don't build for different reasons. `python37` doesn't build because of `ffc`. `python39` doesn't build because of `dijitso`. However, `python38` builds perfectly fine.

Hey, @jonringer and @FRidh, this is a fairly substantial change. I want to run this by both of you before it gets merged. I hope you also have some insight as to why only python37 and python39 fail, but python38 works perfectly.

ravenjoad committed:
dolfin: init at 2019.1.0

remove unneeded comment listing required python packages

dolfin: get licenses correct

dolfin: use with lib

dolfin: correct licenses

ravenjoad committed:
python.pkgs.dijitso: init at dolfin's version

python.pkgs.dijitso: get licenses correct

python.pkgs.dijitso: use with lib

python.pkgs.dijitso: correct licenses

python.pkgs.dijitso: add pythonImportsCheck

ravenjoad committed:
python.pkgs.ffc: init at dolfin's version

python.pkgs.ffc: get licenses correct

python.pkgs.ffc: use with lib

python.pkgs.ffc: correct licenses

python.pkgs.ffc: add pythonImportsCheck

ravenjoad committed:
python.pkgs.fiat: init at dolfin's version

python.pkgs.fiat: get licenses correct

python.pkgs.fiat: correct licenses

ravenjoad committed:
python.pkgs.ufl: init at dolfin's version

python.pkgs.ufl: get licenses correct

python.pkgs.ufl: use with lib

python.pkgs.ufl: correct licenses

python.pkgs.ufl: add pythonImportsCheck

ravenjoad committed:
python.pkgs.fenics: move python fenics bindings to be python module

python.pkgs.fenics: correct pname and sourceRoot

python2 packages were split off

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+  meta = {
+    description = "A domain-specific language for finite element variational forms";
+    homepage = "https://fenicsproject.org/";
+    platforms = lib.platforms.all;
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+    description = "A domain-specific language for finite element variational forms";
+    homepage = "https://fenicsproject.org/";
+    platforms = lib.platforms.all;
+    license = lib.licenses.lgpl3;
```
```suggestion
    license = licenses.lgpl3;
```

lgpl3Only or lgpl3Plus? 

Missing maintainers.

KarlJoad replied:
I have no idea how to maintain this. There wasn't a maintainer for this set of packages before either.

SuperSandro2000 replied:
> There wasn't a maintainer for this set of packages before either.

Which is not great but alright.

KarlJoad replied:
`lgpl3Only` I believe. i may be wrong. Check [here](https://github.com/FEniCS/ufl/blo/3cdfb8acfeeb5404804e12a304b65f58c51d822c/COPYING#L75).

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+    runHook postCheck
+  '';
+
+  meta = {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+    sha256 = "04daxwg4y9c51sdgvwgmlc82nn0fjw7i2vzs15ckdc7dlazmcfi1";
+  };
+
+  propagatedBuildInputs = with python3Packages; [ numpy six ];
+
+  checkInputs = with python3Packages; [ pytest ];
+
+  checkPhase = ''
+    runHook preCheck
+    py.test test/
+    runHook postCheck
+  '';
```
```suggestion
  propagatedBuildInputs = [ numpy six ];

  checkInputs = [ pytestCheckHook ];

  pytestFlagsArray = [ "test/" ];
```

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+, dolfin
+}:
+
+python3Packages.buildPythonPackage rec {
```
```suggestion
buildPythonPackage rec {
```

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+{ stdenv
+, lib
+, fetchurl
+, python3Packages
```
```suggestion
, buildPythonPackage
```

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+, lib
+, fetchurl
+, python3Packages
+, dolfin
```
```suggestion
, dolfin
, numpy
, six
, pytestCheckHook
```

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,33 @@
+  inherit (dolfin) version;
+
+  src = fetchurl {
+    url = "https://bitbucket.org/fenics-project/ufl/downloads/ufl-${version}.tar.gz";
```
Please use fetchFromBitBucket.

SuperSandro2000 commented on pkgs/development/python-modules/fiat/default.nix:
```diff
@@ -0,0 +1,41 @@
+  meta = {
+    description = "Automatic generation of finite element basis functions";
+    homepage = "https://fenicsproject.org/";
+    platforms = lib.platforms.all;
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/fiat/default.nix:
```diff
@@ -0,0 +1,41 @@
+    description = "Automatic generation of finite element basis functions";
+    homepage = "https://fenicsproject.org/";
+    platforms = lib.platforms.all;
+    license = lib.licenses.lgpl3;
```
```suggestion
    license = licenses.lgpl3;
```

lgpl3Only or lgpl3Plus? Missing maintainers.

SuperSandro2000 commented on pkgs/development/python-modules/fiat/default.nix:
```diff
@@ -0,0 +1,41 @@
+    runHook postCheck
+  '';
+
+  meta = {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/development/python-modules/fiat/default.nix:
```diff
@@ -0,0 +1,41 @@
+    rm test/unit/test_fiat.py
+  '';
+
+  checkPhase = ''
```
Please use pytestCheckHook.

SuperSandro2000 commented on pkgs/development/python-modules/fiat/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+  propagatedBuildInputs = with python3Packages; [ numpy six sympy ];
+
+  checkInputs = with python3Packages; [ pytest ];
```
python3Packages is not allowed in python modules. Please use the same solution as shown below.

SuperSandro2000 commented on pkgs/development/python-modules/ffc/default.nix:
```diff
@@ -0,0 +1,52 @@
+, python3Packages
+, dolfin
+}:
+
+with python3Packages;
```
```suggestion
```

same as below

SuperSandro2000 commented on pkgs/development/python-modules/fenics/default.nix:
```diff
@@ -0,0 +1,63 @@
+
+  dontUseCmakeConfigure = true;
+  preConfigure = ''
+    export CMAKE_PREFIX_PATH=${pybind11}/share/cmake/pybind11:$CMAKE_PREFIX_PATH
```
Can we use cmakeFlags for this?

SuperSandro2000 commented on pkgs/development/python-modules/fenics/default.nix:
```diff
@@ -0,0 +1,63 @@
+  meta = {
+    description = "Python bindings for the DOLFIN FEM compiler";
+    homepage = "https://fenicsproject.org/";
+    platforms = lib.platforms.all;
```
```suggestion
```

Set by buildPythonPackage.

SuperSandro2000 commented on pkgs/development/python-modules/dijitso/default.nix:
```diff
@@ -0,0 +1,36 @@
+, dolfin
+}:
+
+python3Packages.buildPythonPackage rec {
```
Same as the other modules.

SuperSandro2000 commented on pkgs/development/libraries/dolfin/default.nix:
```diff
@@ -0,0 +1,107 @@
+
+  installCheckPhase = ''
+    source $out/share/dolfin/dolfin.conf
+    make runtests
```
```suggestion
    make runtests -j$NIX_BUILD_CORES
```

SuperSandro2000 commented on pkgs/development/libraries/dolfin/default.nix:
```diff
@@ -0,0 +1,107 @@
+    pkg-config
+  ];
+
+  buildInputs = [
+    boost
+    python3Packages.dijitso
+    eigen
+    python3Packages.ffc
+    python3Packages.fiat
+    hdf5
+    mpi
+    python3Packages.numpy
+    blas
+    lapack
+    python3Packages.ply
+    python3
+    scotch
+    python3Packages.six
+    python3Packages.sphinx
+    suitesparse
+    swig
+    python3Packages.sympy
+    python3Packages.ufl
+    zlib
+  ];
```
```suggestion
  buildInputs = [
    boost
    eigen
    hdf5
    mpi
    blas
    lapack
    python3
    scotch
    suitesparse
    swig
    zlib
  ] ++ (with python3Packages; [
    dijitso
    ffc
    fiat
    numpy
    ply
    six
    sphinx
    sympy
    ufl
  ]);
```

FRidh commented on pkgs/development/libraries/dolfin/default.nix:
```diff
@@ -0,0 +1,107 @@
+, fetchpatch
+# - Python Dependencies
+, python3
+, python3Packages
```
only `python3` or `python3Packages`, not both.

FRidh commented on pkgs/development/python-modules/ffc/default.nix:
```diff
@@ -0,0 +1,52 @@
+{ stdenv
+, lib
+, fetchurl
+, python3Packages
```
explicitly state required Python packages as parameters

FRidh commented on pkgs/development/python-modules/fiat/default.nix:
```diff
@@ -0,0 +1,41 @@
+{ stdenv
+, lib
+, fetchurl
+, python3Packages
```
explicitly state required Python packages as parameters

jonringer replied:
Using `pythonXPackages` isn't allowed from `python-modules` as this will mix interpreter versions for a given package. For example, if you were to do `python37Packages.<mypkg>`, but the expression pulled from `python3[8]Packages`. Python will be mis-importing several libraries, and this will likely cause issues at runtime.

Please refer to [Python dependency handling](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#handling-dependencies) on the best practices for handling dependencies within `python-modules`. Namely, you should be importing each python package individually, and not refer to another python package set (E.g. `python3Packages`).

If you're new to nixpkgs, you're free to view some of my videos on nixpkgs and python:
- [Python packaging using nixpkgs](https://www.youtube.com/watch?v=jXd-hkP4xnU)
- [Adding a package to nixpkgs unstable](https://www.youtube.com/watch?v=fvj8H5yUKu8)

jonringer commented on pkgs/development/python-modules/ffc/default.nix:
```diff
@@ -0,0 +1,52 @@
+    description = "A compiler for finite element variational forms";
+    homepage = "https://fenicsproject.org/";
+    platforms = lib.platforms.all;
+    license = lib.licenses.lgpl3;
```
Please add yourself as a maintainer when adding a new package.

https://nixos.org/manual/nixpkgs/stable/#reviewing-contributions-new-packages

jonringer commented on pkgs/development/python-modules/ffc/default.nix:
```diff
@@ -0,0 +1,52 @@
+
+  preCheck = ''
+    export HOME=$PWD
+    rm test/unit/ufc/finite_element/test_evaluate.py
```
please add a comment saying way this is necessary

KarlJoad replied:
I have no idea why that is necessary. It was in the original code, and I kept it.

jonringer commented on pkgs/development/python-modules/dijitso/default.nix:
```diff
@@ -0,0 +1,36 @@
+    description = "Distributed just-in-time shared library building";
+    homepage = "https://fenicsproject.org/";
+    platforms = lib.platforms.all;
+    license = lib.licenses.lgpl3;
```
Please add yourself as a maintainer when adding a new package.

https://nixos.org/manual/nixpkgs/stable/#reviewing-contributions-new-packages

KarlJoad replied:
I'm not sure if I should add myself, because this package originally had no maintainers, and I don't think I am qualified enough to handle maintenance of this package. I made this PR because I need the dolfin library for another PR I have open, so I split these modules apart. I have no real idea how they fit together.

jonringer commented on pkgs/development/libraries/dolfin/default.nix:
```diff
@@ -0,0 +1,107 @@
+  meta = {
+    description = "The FEniCS Problem Solving Environment in Python and C++";
+    homepage = "https://fenicsproject.org/";
+    license = lib.licenses.lgpl3;
```
Please add yourself as a maintainer when adding a new package.

https://nixos.org/manual/nixpkgs/stable/#reviewing-contributions-new-packages

ravenjoad committed:
python.pkgs.fenics: explicitly name python deps & pytestCheckHook

python.pkgs.fenics: get licenses correct

ravenjoad committed:
dolfin: move all required python packages to appended with-scoped list

ravenjoad committed:
dolfin: parallelize installcheckPhase

ravenjoad committed:
python.pkgs.fenics: switch fetchurl -> fetchFromBitbucket

ravenjoad committed:
dolfin: switch fetchurl -> fetchFromBitbucket

ravenjoad committed:
python.pkgs.fiat: use disabledTestPaths instead of disabledTests

ravenjoad committed:
python.pkgs.ffc: disable tests that make build fail

ravenjoad committed:
dolfin: disable HDF5 due to build problems

ravenjoad committed:
python.pkgs.fenics: remove unneeded export of CMAKE_PREFIX_PATH

This seems to be unneeded, but I am not sure. I built this package
twice, once with and once without `CMAKE_PREFIX_PATH` set, and it
built fine both times.

SuperSandro2000 commented on pkgs/development/python-modules/fenics/default.nix:
```diff
@@ -0,0 +1,62 @@
+buildPythonPackage rec {
+  pname = "fenics";
+  inherit (dolfin) version;
+  disabled = isPy27;
```
```suggestion
```

python 2 packages got split off.

KarlJoad replied:
Meaning it will no longer be supported and is considered an edge case?

SuperSandro2000 commented on pkgs/development/libraries/dolfin/default.nix:
```diff
@@ -0,0 +1,110 @@
+, fetchpatch
+# - Python Dependencies
+, python3Packages
+# , numpy, sympy, six, sphinx, ufl, ply, dijitso, fiat
```
```suggestion
```

SuperSandro2000 reviewed (commented):
Missing maintainers. Are the licenses lgpl3Only or lgpl3Plus?

KarlJoad commented:
> Missing maintainers. Are the licenses lgpl3Only or lgpl3Plus?

I think we discussed this earlier in the PR. These never had a maintainer to begin with, and I do not have the necessarily knowledge to properly maintain them. I believe @jonringer said that leaving no maintainers for these was OK. If need be, I _guess_, I can be a maintainer, but I do not have a ton of time to do proper maintenance of this.

Still working on licenses. I need to look up the difference between the two to identify what it is.

ravenjoad committed:
python.pkgs.fenics: use with lib

python.pkgs.fenics: correct licenses

SuperSandro2000 reviewed (changes requested):
maintainers are missing.

pythonImportsCheck wouldn't hurt.

KarlJoad commented:
> pythonImportsCheck wouldn't hurt.

I included this in as many of the `init` commits as I could without causing things to break. `fenics` and `fiat` did not work in a `nixpkgs-review` shell when I tried.

SuperSandro2000 commented on pkgs/development/libraries/dolfin/default.nix:
```diff
@@ -0,0 +1,109 @@
+  '';
+
+  meta = with lib; {
+    description = "The FEniCS Problem Solving Environment in Python and C++";
```
Missing maintainer.

SuperSandro2000 commented on pkgs/development/python-modules/fenics/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  meta = with lib; {
+    description = "Python bindings for the DOLFIN FEM compiler";
+    homepage = "https://fenicsproject.org/";
```
Missing maintainer.

SuperSandro2000 commented on pkgs/development/python-modules/ffc/default.nix:
```diff
@@ -0,0 +1,64 @@
+
+  meta = with lib; {
+    description = "A compiler for finite element variational forms";
+    homepage = "https://fenicsproject.org/";
```
Missing maintainer.

SuperSandro2000 commented on pkgs/development/python-modules/fiat/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+  meta = with lib; {
+    description = "Automatic generation of arbitrary order instances of the Lagrange elements on lines, triangles, and tetrahedra";
+    homepage = "https://fenicsproject.org/";
```
Missing maintainer.

SuperSandro2000 commented on pkgs/development/python-modules/ufl/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  meta = with lib; {
+    description = "A domain-specific language for finite element variational forms";
+    homepage = "https://fenicsproject.org/";
```
Missing maintainer.

SuperSandro2000 commented:
Closing as stale
